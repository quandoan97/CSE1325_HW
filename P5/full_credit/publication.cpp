 #include "publication.h"
 #include <string>
 #include <iostream>

 using namespace std;


 bool Publication::is_checked_out() {return checked_out;}
 void Publication::check_out(string name, string phone) {
   if (checked_out) {
     throw Invalid_transaction();
   } else {
     patron = name;
     patron_phone = phone;
     checked_out = true;
   }
 }
   
 void Publication::check_in() {
   if (checked_out) {
     checked_out = false;
   } else {
     throw Invalid_transaction();
   }
 }

 string Publication::to_string() {
   string pub = "\"" + title + "\"" + " by " + author + ", " + copyright + 
     " (" + ages[target_age] + " " + genres[genre] + " " + medias[media] + ") " + 
     "ISBN: " + isbn;
   if (checked_out) {
      pub += "\nChecked out to " + patron + " (" + patron_phone + ")";
   }
   return pub;
 }
