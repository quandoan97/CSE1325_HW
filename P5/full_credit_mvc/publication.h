 #ifndef __PUBLICATION_H
 #define __PUBLICATION_H 201609
 
 #include "media.h"
 #include "genre.h"
 #include "age.h"

 #include <iostream>
 #include <string>

 using namespace std;


 class Publication {
   public:
     Publication(string p_title, 
                 string p_author, 
                 string p_copyright,
                 Genre p_genre, 
                 Media p_media, 
                 Age p_target_age,
                 string p_isbn) : 

                 title(p_title), 
                 author(p_author),
                 copyright(p_copyright), 
                 genre(p_genre), 
                 media(p_media),
                 target_age(p_target_age), 
                 isbn(p_isbn), 
                 patron(""),
                 patron_phone(""),
                 checked_out(false) { }

     bool is_checked_out();

     void check_out(string name, string phone);
     void check_in();

     string to_string();

     // Thrown on check_in if publication isn't checked out
     //   or on cheeck_out if publication is already checked out
     class Invalid_transaction { }; 

   private:
     string title;
     string author;
     string copyright;
     Genre genre;
     Media media;
     Age target_age;
     string isbn;
     string patron;
     string patron_phone;
     bool checked_out;
 };
 #endif
