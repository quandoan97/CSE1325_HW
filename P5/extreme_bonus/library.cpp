#include "library.h"
#include "publication.h"
#include "book.h"
#include "periodical.h"
#include "newspaper.h"
#include "audio.h"
#include "video.h"

void Library::add_publication(Publication* pub) {
  publications.push_back(pub);
}

void Library::add_patron(Patron pat) {
  patrons.push_back(pat);
}

void Library::check_out(int publication_index, int patron_index) {
  publications[publication_index]->check_out(patrons[patron_index]);
}

void Library::check_in(int publication_index) {
  publications[publication_index]->check_in();
}

string Library::publication_to_string(int publication_index) {
  return publications[publication_index]->to_string();
}

string Library::patron_to_string(int patron_index) {
  return patrons[patron_index].to_string();
}

int Library::number_of_publications() {
  return publications.size();
}

int Library::number_of_patrons() {
  return patrons.size();
}

void Library::easter_egg() {
 add_publication(new Book("The Firm", "John Grisham", "1991",
       Genre::fiction, Media::book, Age::adult, "0440245923", 0));
 add_publication(new Book("Foundation", "Isaac Asimov", "1942",
        Genre::fiction, Media::book, Age::adult, "0385177259", 1));
 add_publication(new Book("Foundation and Empire", "Isaac Asimov", "1943",
        Genre::fiction, Media::book, Age::adult, "0385177259", 1));
 add_publication(new Book("Second Foundation", "Isaac Asimov", "1944",
        Genre::fiction, Media::book, Age::adult, "0385177259", 1));
 add_publication(new Audio("War of the Worlds", "Jeff Wayne", "1977",
        Genre::performance, Media::audio, Age::teen, "9780711969148", 0, 1));
 add_publication(new Video("Willy Wonka and the Chocolate Factory", "Roald Dahl", "1971",
        Genre::performance, Media::video, Age::children, "0142410314", 1, 0, 0));
 add_patron(Patron("Larry", "817-555-1111"));
 add_patron(Patron("Curly", "817-555-2222"));
 add_patron(Patron("Moe", "817-555-3333"));
}
