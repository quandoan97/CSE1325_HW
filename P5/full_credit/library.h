 #ifndef __LIBRARY_H
 #define __LIBRARY_H 201609
 #include "publication.h"

 #include <iostream>
 #include <string>
 #include <vector>

 using namespace std;


 class Library {
   public:
     void add_publication(Publication pub);

     void check_out(int publication_index, string patron, string patron_phone);
     void check_in(int publication_index);

     string publication_to_string(int publication_index);

     int number_of_publications();

     void easter_egg();
   private:
     vector<Publication> publications;
 };
 #endif
