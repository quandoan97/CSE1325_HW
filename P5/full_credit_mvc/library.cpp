#include "library.h"

void Library::add_publication(Publication pub) {
  publications.push_back(pub);
}

void Library::check_out(int publication_index, string patron, string patron_phone) {
  publications[publication_index].check_out(patron, patron_phone);
}

void Library::check_in(int publication_index) {
  publications[publication_index].check_in();
}

string Library::publication_to_string(int publication_index) {
  return publications[publication_index].to_string();
}

int Library::number_of_publications() {
  return publications.size();
}

void Library::easter_egg() {
 add_publication(Publication("The Firm", "John Grisham", "1991",
       Genre::fiction, Media::book, Age::adult, "0440245923"));
 add_publication(Publication("Foundation", "Isaac Asimov", "1942",
        Genre::fiction, Media::book, Age::adult, "0385177259"));
 add_publication(Publication("Foundation and Empire", "Isaac Asimov", "1943",
        Genre::fiction, Media::book, Age::adult, "0385177259"));
 add_publication(Publication("Second Foundation", "Isaac Asimov", "1944",
        Genre::fiction, Media::book, Age::adult, "0385177259"));
 add_publication(Publication("War of the Worlds", "Jeff Wayne", "1977",
        Genre::performance, Media::audio, Age::teen, "9780711969148"));
 add_publication(Publication("Willy Wonka and the Chocolate Factory", "Roald Dahl", "1971",
        Genre::performance, Media::video, Age::children, "0142410314"));
}
