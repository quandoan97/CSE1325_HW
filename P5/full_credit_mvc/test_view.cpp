#include "view.h"
#include "publication.h"
#include "library.h"

int main() {
  Library library;
  View view(library);

  bool passed = true;

  Publication foundation("Foundation", "Isaac Asimov", "1942", 
      Genre::fiction, Media::book, Age::adult, "0385177259");
  Publication foundation_empire("Foundation and Empire", "Isaac Asimov", "1943", 
      Genre::fiction, Media::book, Age::adult, "0385177259");
  Publication foundation2("Second Foundation", "Isaac Asimov", "1944", 
      Genre::fiction, Media::book, Age::adult, "0385177259");
  Publication war("War of the Worlds", "Jeff Wayne", "1977",
      Genre::performance, Media::audio, Age::teen, "9780711969148");
  Publication wonka("Willy Wonka and the Chocolate Factory", "Roald Dahl", "1971",
      Genre::performance, Media::video, Age::children, "0142410314");

  library.add_publication(foundation);
  library.add_publication(foundation_empire);
  library.add_publication(foundation2);
  library.add_publication(war);
  library.add_publication(wonka);

  library.check_out(1, "Curly", "817-555-1212");

  //
  // Test menu
  //
  cout << endl;
  view.show_menu();

  //
  // Test list publications
  //
  cout << endl;
  view.list_publications();

  //
  // Show results
  //
  cout << endl << "Verify output for pass or fail" << endl;
}
