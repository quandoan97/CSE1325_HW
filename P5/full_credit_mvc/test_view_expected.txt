

=================================
CSE1325 Library Management System
=================================

Publications
------------
(1) Add publication
(2) List all publications
(3) Check out publication
(4) Check in publication

Utility
-------
(9) Help
(0) Exit



----------------------------
List of Library Publications
----------------------------
0) "Foundation" by Isaac Asimov, 1942 (adult fiction book) ISBN: 0385177259
1) "Foundation and Empire" by Isaac Asimov, 1943 (adult fiction book) ISBN: 0385177259
Checked out to Curly (817-555-1212)
2) "Second Foundation" by Isaac Asimov, 1944 (adult fiction book) ISBN: 0385177259
3) "War of the Worlds" by Jeff Wayne, 1977 (teen performance audio) ISBN: 9780711969148
4) "Willy Wonka and the Chocolate Factory" by Roald Dahl, 1971 (children performance video) ISBN: 0142410314

Verify output for pass or fail
